content/articles/2017-12-testthat-2-0.Rmd

```diff
@@ -0,0 +1,207 @@
+    needs to be used in conjunction with git. It replaces `expect_output_file()`
+    which is now soft-deprecated.
+
+    ```{r, error = TRUE}
+    # File to save results: would usually be filename, which will be stored 
+    # in tests/testthat.
+    tmp <- tempfile()
+    
+    # The first run always succeeds, but warns
+    expect_known_output(mtcars[1:10, ], tmp, print = TRUE)
+    
+    # Subsequent runs will suceed only if the file is unchanged
+    # This will succeed:
+    expect_known_output(mtcars[1:10, ], tmp, print = TRUE)
+    
+    # This will fail
+    expect_known_output(mtcars[1:9, ], tmp, print = TRUE)
+    ```
+
+* `expect_known_value()` replaces `expect_equal_to_reference()`, which
+  has been soft-deprecated. It gains an update argument defaulting to `TRUE`.
+  This changes behaviour from the previous version, and soft-deprecated
+  `expect_equal_to_reference()` gets `update = FALSE`. 
+
+We've also improved tools for failures:
+
+* `expect_condition()` works like `expect_error()` but captures any
+  condition, not just error conditions.
+
+*  `expect_error()` gains a `class` argument that allows you to make an 
+  assertion about the class of the error object (#530).
+
+We've also added `expect_setequal()` to check if that compares two sets (stored in vectors), ignoring duplicates and differences in order. Finally we added few new helpers for skipping tests:
+
+* `skip_if()` makes it easy to skip a test when a condition is true.
+  For example, use `skip_if(getRversion() <= 3.1)` to skip a test in older
+  R versions.
+
+* `skip_if_translated()` skips tests if you're running in an locale
+  where translations are likely to occur. Use this to avoid spurious failures 
+  when checking the text of error messages in non-English locales.
+  
+* `skip_if_not_installed()` gains new `minimum_version` argument. This allows
+  you to only test if a minimum version requrement is met, e.g.
+  
+    ```{r}
+    skip_if_not_installed("ggplot2", "2.0.0")
+    ```
+
+## Quasiquotation support
+
+All expectations can now use [unquoting](https://adv-r.hadley.nz/tidy-eval) with `!!`. This makes it much easier to generate informative failure messages when running tests in a for loop or function. For example take this test:
+
+```{r, error = TRUE}
+f <- function(i) if (i > 3) i * 9 else i * 10
+
+for (i in 1:5) {
+  expect_equal(f(i), i * 10)
+}
+```
+
+The error message is not great because you don't know which iteration caused the problem! You can resolve that problem by using the unquoting operator `!!` (pronounced bang-bang):
+
+```{r, error = TRUE}
+for (i in 1:5) {
+  expect_equal(f(!!i), !!(i * 10))
+}
+```
+
+(Note that this is not tidy evaluation per se, but is closely related. This works a little differently in testthat to tidyverse package because the only quoting involved is for error message. See `?quasi_label()` for more details. At this time you can not unquote quosures.)
```

Comment:
I italicized *per se*, "vs. tidyverse packageS", "is for THE error message":

> (Note that this is not tidy evaluation *per se*, but is closely related. This works a little differently in testthat vs. tidyverse packages, because the only quoting involved is for the error message. See `?quasi_label()` for more details. At this time you can not unquote quosures.)